### DISCUSSION TITLE
Understanding gittargets

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/gittargets#15] [Q&A] [answered]
I have been playing with the example of the vignette, but I don't grasp how gittargets tracks the data status (and revert it when needed): I expected to see a commit specific for each `tar_git_snapshot()`. 

It seems like it uses a git submodule or something similar because when I used `tar_git_snapshot` without committing all the code I got an error:

```
• The new snapshot will be a data commit that maps to the following code commit:
  dff058c1631ea847fb03e8e6b2b18233f5df3dc2
  Switch to UKgas dataset
• Please make sure the code repo and targets pipeline are clean and up to date.

1: yes
2: no

Selection: 1
• Data snapshot already exists for the current code commit.
• The new data snapshot will supersede the old one in data branch:
  code=dff058c1631ea847fb03e8e6b2b18233f5df3dc2
Error in `processx::run(command = tar_git_binary(), args = args, wd = repo, …`:
! System command 'git.exe' failed
---
Exit status: 1
Stderr:
error: Your local changes to the following files would be overwritten by checkout:
	.gittargets
	meta/meta
	meta/process
	objects/data
	objects/result
Please commit your changes or stash them before you switch branches.
Aborting
```

Exploring via the terminal under the _targets folder I see:

```
$ git status
On branch code=cf3a3ec40baba84821b696cf7d98a3e2f6402b29
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .gitignore

nothing added to commit but untracked files present (use "git add" to track)
```

Can I combine using git from the terminal and the `tar_git_checkout()`? If I'm understanding well all the process tar_git_checkout commits the data (not the code). How should I run the same process via the terminal (i.e. `git add _targets/* ...` ?

I see there is a workaround related to .gittatributes

**Comment by wlandau** [comment, marked as answer]
> I have been playing with the example of the vignette, but I don't grasp how gittargets tracks the data status (and revert it when needed): I expected to see a commit specific for each tar_git_snapshot().

The snapshot model is explained at https://docs.ropensci.org/gittargets/articles/git.html#snapshot-model. Each snapshot is a commit, those commits are partitioned into branches, and each branch corresponds to a unique code commit. In other words, `gittargets` uses branching to group the data commits by the corresponding code commits. This is what allows it to revert to a data commit when you revert to an old code commit.

> Can I combine using git from the terminal and the tar_git_checkout()? 

You could in theory, but it is much easier to check out the code commit you want and then call `tar_git_checkout()` with no arguments. If a checkout fails, it could be that your code workspace or data workspace has uncommitted changes. Before reverting to either an old code commit or an old data commit, I highly recommend making sure everything is up to date (committed with nothing unstaged or staged in the workspace). This is a common difficulty with using command line Git as well.
